src/bucket.rs

```diff
@@ -49,20 +49,21 @@ pub struct Bucket {
 
 impl Bucket {
     /// Construct a new S3 bucket
-    pub fn new(endpoint: Url, path_style: bool, name: String, region: String) -> Option<Self> {
+    // 这里可以有 pr
+    pub fn new<S: AsRef<str>>(endpoint: Url, path_style: bool, name: S, region: S) -> Option<Self> {
```

Comment:
Since these get converted to `String` it would make sense to take `ToString` or `Into<String>`. Also I'd go for a different generic type between `name` and `region`, since they could originate from different sources with different types.

Reply:
Yeah, your opinion makes sense. I will think about it.  
It's 4:00 in my time zone now, I need to sleep. Tomorrow I will back to discuss this implementation.  
Thanks for your patience.

Reply:
Hello, I'm back.  
In my opinion, the role of Bucket::new is to constructt a new S3 bucket, the `name` and `region` will be converted to String In the end. So the problem is when we convert them.  
The old implementation requested the API user to specify the two parameters as `String`, while the new implementation allow the user  use `String` or `&str` or any other types which implement `AsRef<str>`, and the job of conversion is placed inside Bucket::new function.  
So I think the advantage of the new implementation is increasing flexibility of the code, even if it brings some additional cost of runtime. Compared to the little additional cost, I prefer the greater flexibility.

Reply:
`Into<String>` would be able to achieve the same thing as `AsRef<str>` though, but with the advantage of not allocating if the generic type were to be a `String`.

Here's a demo to demonstrate it:

* [taking `AsRef<str>`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c6d477595c2ae803c90898c53ecd3218)
* [taking `Into<String>`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6e1cdbb59b6fb292fddced7f6074fba4)

Both cases are able to accept both `String` and `&str`, but the second one is better because it doesn't make a new allocation if the type is a `String`

Reply:
Cool! I have fixed it. Thanks for your suggestions and patience!